### PR TITLE
[Python CLI] Secure profile secret handling and config storage

### LINKED ISSUE #4769: CLI profile commands expose client secrets in plaintext
## Problem

The CLI currently handles `client_secret` values insecurely in profile management commands:

* `profiles create` and `profiles update` read secrets using `input()`, causing the secret to be visible on-screen and potentially stored in terminal scrollback/history.
* `profiles get` prints the complete profile configuration, including the raw `client_secret`.
* The Polaris configuration file is written using default file permissions, which may allow unintended access by other users on the same system.

Affected locations:

* `client/python/apache_polaris/cli/command/profiles.py`

  * Lines 66–70
  * Lines 75–76
  * Lines 124–126
  * Lines 168–172

## Expected Behavior

* Secret values should be entered without echoing to the terminal.
* Displayed profile information should mask sensitive fields such as `client_secret`.
* Configuration files containing secrets should be written with user-only permissions (0600).

## Proposed Solution

1. Replace `input()` with `getpass.getpass()` when reading `client_secret`.
2. Introduce a helper (e.g., `format_profile_for_display()`) that masks sensitive fields before printing profile information.
3. Write `~/.polaris/.polaris.json` using an atomic write pattern:

   * Create a temporary file.
   * Set permissions to `0o600`.
   * Replace the existing file using `os.replace()`.

## Impact

This change improves credential security while preserving the existing configuration schema and CLI behavior.

## Additional Context

The issue affects local credential handling and may expose secrets through terminal output, logs, screen recordings, shared machines, or improperly permissioned configuration files.

### To Reproduce

_No response_

### Actual Behavior

_No response_

### Expected Behavior

_No response_

### Additional context

_No response_

### System information

_No response_

### PR DESCRIPTION
**Pull request by venkateshwaracholan**
Fixes #4769

## Summary
Hardens Python CLI profile secret handling:
- Fully mask `client_secret` in profile display output (`profiles get`)
- Use plain `input()` for secret entry during create/update (aligned with AWS CLI workflow)
- Use a static update prompt: `Enter Polaris Client Secret (empty to reuse previous value)`
- Use atomic writes via temporary file + `os.replace()`
- Centralize config I/O in `profile_config.py` with read-only `load_profiles()`

## Testing
- Secret masking (including `None` and empty string edge cases)
- Secure config permissions on save (including correcting permissive existing files)
- Atomic writes with no leftover temp files
- `ensure_config_file_permissions` no-op when config file is missing
- Profile command integration tests for create/update/get flows

## Checklist
- [ ] 🛡️ Don't disclose security issues! (contact security@apache.org)
- [ ] 🔗 Clearly explained why the changes are needed, or linked related issues: Fixes #
- [ ] 🧪 Added/updated tests with good coverage, or manually tested (and explained how)
- [ ] 💡 Added comments for complex logic
- [ ] 🧾 Updated `CHANGELOG.md` (if needed)
- [ ] 📚 Updated documentation in `site/content/in-dev/unreleased` (if needed)

### CODE REVIEW

File: client/python/apache_polaris/cli/profile_config.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+CONFIG_FILE_MODE = 0o600
+
+
+def mask_client_secret(client_secret: str | None) -> str | None:
+    if client_secret is None:
+        return None
+    if len(client_secret) <= 4:
+        return "*" * len(client_secret)
+    return "*" * (len(client_secret) - 4) + client_secret[-4:]
```
Comment (dimas-b) [inline client/python/apache_polaris/cli/profile_config.py]: Why show any part of a secret value?

Comment (venkateshwaracholan) [reply client/python/apache_polaris/cli/profile_config.py]: I originally exposed the last few characters as a usability cue so users could distinguish between stored secrets without revealing the full value.

I've updated the implementation to fully mask the client secret in displayed output instead.

Comment (dimas-b) [reply client/python/apache_polaris/cli/profile_config.py]: Well, users can always read the config file if they need to double check the value :wink:

File: client/python/apache_polaris/cli/command/profiles.py [hunk 2]
```diff
@@ -117,12 +116,15 @@ def _update_profile(self, name: str) -> None:
...
+            masked_secret = format_profile_for_display(profiles[name]).get(
+                "client_secret"
+            )
 
             client_id = (
                 input(f"Polaris Client ID [{current_client_id}]: ") or current_client_id
             )
             client_secret = (
-                input(f"Polaris Client Secret [{current_client_secret}]: ")
+                getpass(f"Polaris Client Secret [{masked_secret}]: ")
```
Comment (dimas-b) [inline client/python/apache_polaris/cli/command/profiles.py]: `masked_secret` always has the same value (logically), could we use a simpler (static) message here? e.g. `Enter Polaris Client Secret (empty to reuse previous value)`?

**Review by dimas-b — COMMENTED** [review]
LGTM overall... I'll defer to @MonkeyCanCode 's review for Python specifics 🙂

**Comment by MonkeyCanCode** [comment]
> LGTM overall... I'll defer to @MonkeyCanCode 's review for Python specifics 🙂 

Thanks for tagging. I will take a look later tonight.

**Comment by MonkeyCanCode** [comment]
One thing I am not sure about is does it really makes sense to use getpass. For example, the reason on why I didn't do so earlier was to matching to how `aws configure` works:
```
➜  ~ aws configure
AWS Access Key ID [None]: user
AWS Secret Access Key [None]: pass
Default region name [None]:
```
With aws cli, the secret access key is plain text when you tried to create one. Then when tried to access, it will show last 4 characters instead of full `*`:
```
➜  ~ aws configure list
      Name                    Value             Type    Location
      ----                    -----             ----    --------
   profile                <not set>             None    None
access_key     ****************user shared-credentials-file
secret_key     ****************pass shared-credentials-file
```

This is more preference IMO. Both will work but do we want to match to how AWS cli is handling those or we can go with our own implementation.

**Comment by venkateshwaracholan** [comment]
One difference I was thinking about is the user workflow.
With AWS, credentials are something many users interact with regularly across multiple accounts, profiles, and tools, so matching established AWS CLI conventions has a lot of value.

For Polaris, my impression is that the client secret is usually entered once when a profile is created and then rarely revisited. In that workflow, hiding the secret during entry feels a bit more natural since users are typically copy/pasting a generated value rather than typing and managing credentials day to day.

Curious to hear both your thoughts on that perspective, @MonkeyCanCode and @dimas-b.

**Comment by MonkeyCanCode** [comment]
> One difference I was thinking about is the user workflow.
> 
> With AWS, credentials are something many users interact with regularly across multiple accounts, profiles, and tools, so matching established AWS CLI conventions has a lot of value.
> 
> 
> 
> For Polaris, my impression is that the client secret is usually entered once when a profile is created and then rarely revisited. In that workflow, hiding the secret during entry feels a bit more natural since users are typically copy/pasting a generated value rather than typing and managing credentials day to day.
> 
> 
> 
> Curious to hear both your thoughts on that perspective, @MonkeyCanCode and @dimas-b.
> 
> 

IMO, it should be similar. Take an example, assuming I have 3 AWS accounts (dev, qa, and prod), I would create 3 profiles and one for each. Then when using AWS CLI, I would use '--profile xxx' depends on which env I am working with. In this case, I would need to create 3 profiles (which CLI will show plain text when setting up profiles).

Now assuming we do the same for Polaris root account per AWS env, I would expected to perform the same setup profile 3 times (one per env).

In this case, the profile is only created once per environment. This can be handy imo when we are creating new principal where a random client id and client secret will be print to screen. In this case, I would think it is cleaner to show plain text when setting up profile for this new principal (as I not 100% sure what I have in my clipboard...I could paste to another terminal to verify but extra step IMO).

**Comment by venkateshwaracholan** [comment]
Thanks @MonkeyCanCode That makes sense, especially around the multi-environment profile setup workflow and being able to verify a pasted secret after `principals create`.

I've updated the PR to use plain `input()` for secret entry to align with the AWS CLI behavior you described, and switched the update prompt to a static message based on @dimas-b's suggestion.

The masking, config permission, and atomic write changes remain unchanged.

**Comment by MonkeyCanCode** [comment]
> Thanks @MonkeyCanCode That makes sense, especially around the multi-environment profile setup workflow and being able to verify a pasted secret after `principals create`.
> 
> 
> 
> I've updated the PR to use plain `input()` for secret entry to align with the AWS CLI behavior you described, and switched the update prompt to a static message based on @dimas-b's suggestion.
> 
> 
> 
> The masking, config permission, and atomic write changes remain unchanged.
> 
> 

Thank. I will take a look tomorrow.

File: client/python/apache_polaris/cli/command/profiles.py [hunk 3]
```diff
@@ -59,15 +61,11 @@ class ProfilesCommand(Command):
     def _load_profiles(self) -> Dict[str, Dict[str, Any]]:
         if not os.path.exists(CONFIG_FILE):
```
Comment (MonkeyCanCode) [inline client/python/apache_polaris/cli/command/profiles.py]: With the introduction of `profile_config.py`, this file exists check is no longer needed as `load_profiles()` also perform the same.

File: client/python/apache_polaris/cli/profile_config.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        return
+    current_mode = stat.S_IMODE(os.stat(path).st_mode)
+    if current_mode != CONFIG_FILE_MODE:
+        os.chmod(path, CONFIG_FILE_MODE)
+
+
+def load_profiles() -> Dict[str, Dict[str, Any]]:
+    if not os.path.exists(CONFIG_FILE):
+        return {}
+    ensure_config_file_permissions()
```
Comment (MonkeyCanCode) [inline client/python/apache_polaris/cli/profile_config.py]: It is not idea to change file permission here as the load operation should be read-only.

File: client/python/tests/test_profile_config.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+# under the License.
+#
+
+import json
+import os
+import stat
+import tempfile
+import unittest
+from unittest.mock import patch
+
```
Comment (MonkeyCanCode) [inline client/python/tests/test_profile_config.py:26]: Consider adding the missing tests to handle edge cases such as:
1.  mask_client_secret(None)
2.  mask_client_secret("")
3.  ensure_config_file_permissions on non-existed file

**Review by MonkeyCanCode — COMMENTED** [review]
Thanks for the contribution. Please also update PR description to reflect the current state as the existed one is outdated (e.g. we no longer use getpass here now)

**Comment by venkateshwaracholan** [comment]
@MonkeyCanCode Thanks for the review.

I've addressed the feedback:

Updated the PR description to reflect the current implementation.
Removed the duplicate file existence checks.
Made load_profiles() read-only.
Added the requested edge-case tests for secret masking and missing config files.

The latest changes are in commit f4851b7

File: client/python/apache_polaris/cli/profile_config.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import tempfile
+from typing import Any, Dict
+
+from apache_polaris.cli.constants import CONFIG_DIR, CONFIG_FILE
+
+CONFIG_FILE_MODE = 0o600
+MASKED_CLIENT_SECRET = "********"
+
+
+def mask_client_secret(client_secret: str | None) -> str | None:
```
Comment (MonkeyCanCode) [inline client/python/apache_polaris/cli/profile_config.py]: nit: most of the code in this project is using `Optional[str]` instead.

File: client/python/apache_polaris/cli/profile_config.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            json.dump(profiles, f, indent=2)
+            f.flush()
+            os.fsync(f.fileno())
+        os.chmod(temp_path, CONFIG_FILE_MODE)
+        os.replace(temp_path, CONFIG_FILE)
+        temp_path = None
+    finally:
+        if temp_path is not None and os.path.exists(temp_path):
+            os.unlink(temp_path)
+    ensure_config_file_permissions()
```
Comment (MonkeyCanCode) [inline client/python/apache_polaris/cli/profile_config.py]: As we are doing `os.replace` (https://docs.python.org/3/library/os.html#os.replace) above, this line is no-ops as permission is already set to CONFIG_FILE_MODE.

**Comment by MonkeyCanCode** [comment]
@venkateshwaracholan I took another look and overall looks good now. I left some comments. Please take a look and thanks for contribution.

**Comment by venkateshwaracholan** [comment]
@MonkeyCanCode Thanks for the review.
- Switched `str | None` to `Optional[str]` for consistency with the existing codebase.
- Removed the redundant `ensure_config_file_permissions()` call after `os.replace()`, since the file permissions are already set on the temporary file before replacement.
- Updated the related tests accordingly.

Thanks for taking another look.

File: client/python/apache_polaris/cli/command/profiles.py [hunk 8]
```diff
@@ -56,16 +56,10 @@ class ProfilesCommand(Command):
     profile_name: Optional[str] = None
 
     def _load_profiles(self) -> Dict[str, Dict[str, Any]]:
```
Comment (MonkeyCanCode) [inline client/python/apache_polaris/cli/command/profiles.py]: as now the _load_profiles and _save_profiles are direct wrapper on top of load_profiles and save_profiles, do we still need the helper function here?

File: client/python/apache_polaris/cli/profile_config.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import tempfile
+from typing import Any, Dict, Optional
+
+from apache_polaris.cli.constants import CONFIG_DIR, CONFIG_FILE
+
+CONFIG_FILE_MODE = 0o600
+MASKED_CLIENT_SECRET = "********"
+
+
+def mask_client_secret(client_secret: Optional[str]) -> Optional[str]:
```
Comment (MonkeyCanCode) [inline client/python/apache_polaris/cli/profile_config.py]: wondering why do we want to do a check here? should we not just blindly use MASKED_CLIENT_SECRET here regardless the value of client_secret?

**Comment by MonkeyCanCode** [comment]
> @MonkeyCanCode Thanks for the review.
> 
> * Switched `str | None` to `Optional[str]` for consistency with the existing codebase.
> * Removed the redundant `ensure_config_file_permissions()` call after `os.replace()`, since the file permissions are already set on the temporary file before replacement.
> * Updated the related tests accordingly.
> 
> Thanks for taking another look.

hi @venkateshwaracholan , thanks for the fix. Two more things which I just noticed after recent refactor and would like to know more about the decision behinds.

**Comment by venkateshwaracholan** [comment]
@MonkeyCanCode Thanks for taking another look!

I've addressed both comments:

* Removed the `_load_profiles()` / `_save_profiles()` wrappers since they were just pass-through methods after the refactor, so calling `load_profiles()` / `save_profiles()` directly keeps things simpler.
* Simplified the secret display logic so any profile with a `client_secret` field now always displays `MASKED_CLIENT_SECRET`, regardless of the stored value. I also updated the associated tests to reflect this behavior.

Thanks again for the feedback!

**Review by MonkeyCanCode — APPROVED** [review]

**Comment by MonkeyCanCode** [comment]
LGTM. Thanks @venkateshwaracholan
